providers/implementations/kdfs/argon2.c

```diff
@@ -0,0 +1,1488 @@
+                if (l >= ctx->threads)
+                    if (crypto_thread_join(ctx->openssl_ctx, t[l - ctx->threads],
+                                           NULL) == 0)
+                        goto fail;
+
+                p.pass = r;
+                p.lane = l;
+                p.slice = (uint8_t)s;
+                p.index = 0;
+
+                t_data[l].ctx = ctx;
+                memcpy(&(t_data[l].pos), &p, sizeof(ARGON2_POS));
+                t[l] = crypto_thread_start(ctx->openssl_ctx, &fill_segment_thr,
+                                           (void*) &t_data[l]);
+                if (t[l] == NULL) {
+                    for (ll = 0; ll < l; ++ll)
+                        crypto_thread_join(ctx->openssl_ctx, t[ll], NULL);
+                    t[ll] = NULL;
+                    goto fail;
+                }
+            }
+            for (l = ctx->lanes - ctx->threads; l < ctx->lanes; ++l) {
+                if (crypto_thread_join(ctx->openssl_ctx, t[l], NULL) == 0)
+                    goto fail;
+                t[l] = NULL;
+            }
+        }
+    }
+
+    crypto_thread_clean(ctx->openssl_ctx, NULL);
+
+    OPENSSL_free(t_data);
+    OPENSSL_free(t);
+
+    return 1;
+
+fail:
+    if (t_data != NULL)
+        OPENSSL_free(t_data);
+    if (t != NULL)
+        OPENSSL_free(t);
+    return 0;
+}
+
+#endif /* ARGON2_NO_THREADS OPENSSL_THREADS */
+
+static int fill_mem_blocks_st(ARGON2_CTX *ctx)
+{
+    uint32_t r, s, l;
+
+    for (r = 0; r < ctx->passes; ++r)
+        for (s = 0; s < ARGON2_SYNC_POINTS; ++s)
+            for (l = 0; l < ctx->lanes; ++l)
+                fill_segment(ctx, r, l, s, 0);
+    return 1;
+}
+
+static ossl_inline int fill_memory_blocks(ARGON2_CTX *ctx)
+{
+    return ctx->threads == 1 ? fill_mem_blocks_st(ctx) : fill_mem_blocks_mt(ctx);
+}
+
+static int validate_inputs(const ARGON2_CTX *ctx)
+{
+    /*
+     * due to -Werror=type-limits, some of the comparisons are made as
+     * X+1 > Y+1 rather than X > Y, as this caused problems when some
+     * lower limits were set to zero.
+     */
+
+    if (ctx == NULL)
+        return 0;
+
+    switch (ctx->type) {
+    case ARGON2_D:
+    case ARGON2_I:
+    case ARGON2_ID:
+        break;
+    default:
+        return 0;
+    }
+
+    if (ctx->out == NULL)
+        return 0;
+    if (ctx->outlen < ARGON2_MIN_OUTLEN)
+        return 0;
+    if (ctx->outlen > ARGON2_MAX_OUTLEN)
+        return 0;
+    if (ctx->pwd == NULL && ctx->pwdlen != 0)
+        return 0;
+    if (ctx->pwdlen+1 < ARGON2_MIN_PWD_LENGTH+1)
+        return 0;
+    if (ctx->pwdlen > ARGON2_MAX_PWD_LENGTH)
+        return 0;
+    if (ctx->salt == NULL && ctx->saltlen != 0)
+        return 0;
+    if (ctx->saltlen < ARGON2_MIN_SALT_LENGTH)
+        return 0;
+    if (ctx->saltlen > ARGON2_MAX_SALT_LENGTH)
+        return 0;
+    if (ctx->secret == NULL && ctx->secretlen != 0)
+        return 0;
+    if (ctx->secret != NULL && ARGON2_MIN_SECRET+1 > ctx->secretlen+1)
```

Comment:
Not sure I have seen this +1 stuff anywhere else in the code.

Reply:
This X+1 > Y+1 was used to appease some overzealous Travis checks that would've triggered a warning otherwise.